network/hybridNetwork.go

```diff
@@ -0,0 +1,233 @@
+// Copyright (C) 2019-2024 Algorand, Inc.
+// This file is part of go-algorand
+//
+// go-algorand is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// go-algorand is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with go-algorand.  If not, see <https://www.gnu.org/licenses/>.
+
+package network
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"sync"
+
+	"github.com/algorand/go-algorand/config"
+	"github.com/algorand/go-algorand/logging"
+	"github.com/algorand/go-algorand/network/addr"
+	"github.com/algorand/go-algorand/protocol"
+)
+
+// HybridP2PNetwork runs both P2PNetwork and WebsocketNetwork to implement the GossipNode interface
+type HybridP2PNetwork struct {
+	p2pNetwork *P2PNetwork
+	wsNetwork  *WebsocketNetwork
+	genesisID  string
+
+	useP2PAddress bool
+}
+
+// NewHybridP2PNetwork constructs a GossipNode that combines P2PNetwork and WebsocketNetwork
+func NewHybridP2PNetwork(log logging.Logger, cfg config.Local, datadir string, phonebookAddresses []string, genesisID string, networkID protocol.NetworkID, nodeInfo NodeInfo) (*HybridP2PNetwork, error) {
+	// supply alternate NetAddress for P2P network
+	p2pcfg := cfg
+	p2pcfg.NetAddress = cfg.P2PListenAddress
+	p2pnet, err := NewP2PNetwork(log, p2pcfg, datadir, phonebookAddresses, genesisID, networkID, nodeInfo)
+	if err != nil {
+		return nil, err
+	}
+	wsnet, err := NewWebsocketNetwork(log, cfg, phonebookAddresses, genesisID, networkID, nodeInfo, p2pnet.PeerID(), p2pnet.PeerIDSigner())
+	if err != nil {
+		return nil, err
+	}
+	return &HybridP2PNetwork{
+		p2pNetwork: p2pnet,
+		wsNetwork:  wsnet,
+		genesisID:  genesisID,
+	}, nil
+}
+
+// Address implements GossipNode
+func (n *HybridP2PNetwork) Address() (string, bool) {
+	// TODO map from configuration? used for REST API, goal status, algod.net, etc
+	if n.useP2PAddress {
+		return n.p2pNetwork.Address()
+	}
+	return n.wsNetwork.Address()
+}
+
+type hybridNetworkError struct{ p2pErr, wsErr error }
+
+func (e *hybridNetworkError) Error() string {
+	return fmt.Sprintf("p2pErr: %s, wsErr: %s", e.p2pErr, e.wsErr)
+}
+func (e *hybridNetworkError) Unwrap() []error { return []error{e.p2pErr, e.wsErr} }
+
+func (n *HybridP2PNetwork) runParallel(fn func(net GossipNode) error) error {
```

Comment:
I like the `runParallel` approach, almost everything here is fire and forget.

Reply:
Are you comfortable with the approach as-is or think we should block merging to change it?